pkg/sfu/downtrack.go

```diff
@@ -289,85 +303,66 @@ func (d *DownTrack) SetTransceiver(transceiver *webrtc.RTPTransceiver) {
 	d.transceiver = transceiver
 }
 
-func (d *DownTrack) maybeTranslateVP8(pkt *rtp.Packet, meta packetMeta) error {
-	if d.vp8Munger == nil || len(pkt.Payload) == 0 {
+// WriteRTP writes a RTP Packet to the DownTrack
+func (d *DownTrack) WriteRTP(extPkt *buffer.ExtPacket, layer int32) error {
+	d.lastRTP.set(time.Now().UnixNano())
+
+	if !d.bound.get() {
 		return nil
 	}
 
-	var incomingVP8 buffer.VP8
-	if err := incomingVP8.Unmarshal(pkt.Payload); err != nil {
-		return err
+	tp, err := d.getTranslationParams(extPkt, layer)
+	if tp.shouldSendPLI {
+		d.lastPli.set(time.Now().UnixNano())
+		d.receiver.SendPLI(layer)
 	}
-
-	translatedVP8 := meta.unpackVP8()
-	payload, err := d.translateVP8Packet(pkt, &incomingVP8, translatedVP8, false)
-	if err != nil {
+	if tp.shouldDrop {
+		d.pktsDropped.add(1)
 		return err
 	}
 
-	pkt.Payload = payload
-	return nil
-}
-
-// Writes RTP header extensions of track
-func (d *DownTrack) writeRTPHeaderExtensions(hdr *rtp.Header) error {
-	// clear out extensions that may have been in the forwarded header
-	hdr.Extension = false
-	hdr.ExtensionProfile = 0
-	hdr.Extensions = []rtp.Extension{}
-
-	for _, ext := range d.rtpHeaderExtensions {
-		if ext.URI != sdp.ABSSendTimeURI {
-			// supporting only abs-send-time
-			continue
-		}
-
-		sendTime := rtp.NewAbsSendTimeExtension(time.Now())
-		b, err := sendTime.Marshal()
+	payload := extPkt.Packet.Payload
+	if tp.vp8 != nil && tp.vp8.header != nil {
+		incomingVP8, _ := extPkt.Payload.(buffer.VP8)
+		payload, err = d.translateVP8Packet(&extPkt.Packet, &incomingVP8, tp.vp8.header)
 		if err != nil {
+			d.pktsDropped.add(1)
 			return err
 		}
+	}
 
-		err = hdr.SetExtension(uint8(ext.ID), b)
-		if err != nil {
-			return err
+	if d.sequencer != nil {
+		meta := d.sequencer.push(extPkt.Packet.SequenceNumber, tp.rtp.sequenceNumber, tp.rtp.timestamp, 0, extPkt.Head)
+		if meta != nil && tp.vp8 != nil && tp.vp8.header != nil {
```

Comment:
do we need to check vp8.header here? looks like header always has value if tp.vp8 is not nil